nav2_controller/src/controller_server.cpp

```diff
@@ -382,8 +426,16 @@ void ControllerServer::computeControl()
       throw nav2_core::ControllerException("Failed to find goal checker name: " + gc_name);
     }
 
+    std::string pc_name = action_server_->get_current_goal()->progress_checker_id;
+    std::string current_progress_checker;
+    if (findProgressCheckerId(pc_name, current_progress_checker)) {
+      current_progress_checker_ = current_progress_checker;
+    } else {
+      throw nav2_core::ControllerException("Failed to find progress checker name: " + pc_name);
+    }
+
     setPlannerPath(action_server_->get_current_goal()->path);
-    progress_checker_->reset();
+    progress_checkers_[current_progress_checker_]->reset();
```

Comment:
Shouldn't you reset all of them? How do you synchronize state that they have when switching between them within a single navigation request? That doesn't seem like a crazy thing to do / need if you're having multiple progress checkers for a single navigation task

Reply:
What about preemption in `updateGlobalPath`? You don't change the progress checker to the new request's value. Is that intentional? I suppose that would handle the request synchronization based isuses I describe above

Reply:
> What about preemption in `updateGlobalPath`? You don't change the progress checker to the new request's value. Is that intentional? I suppose that would handle the request synchronization based isuses I describe above. But you'd need to be explicit about that and check if there's been a change to reject + log the invalid preemption request. Or, handle the switching state between them.

My bad, I missed it.
I see 3 options:

1. rejecting the goal in case of a switch of progress checker during a preempt
2. resetting the new progress checker in case of a switch.
3. some form of synchronization but I would need to think more about it

Reply:
> rejecting the goal in case of a switch of progress checker during a preempt

That's kind of kicking the can down the road, but I'm OK with that

> resetting the new progress checker in case of a switch.

I think that's the best solution. It supports preemption and hopefully a user knows that if they change the checkers the state will be reset. I don't foresee a situation where someone's doing that alot (... but now that I say that ...) and even if replanning comes in at a regular rate from the BT, as long as it doesn't change the progress checker in the request, we just don't reset it so it should be fine.

---

I leave it to you among the two. I think 2 is the best option but I don't feel strongly about it.